dashboard/app/assets/javascripts/client_state.js

```diff
@@ -42,6 +42,43 @@ dashboard.clientState.reset = function() {
 };
 
 /**
+ * Returns the client-cached copy of the level source for the given
+ * scriptLevelId, if it's newer than the given timestamp.
+ * @param {number} scriptLevelId
+ * @param {number} timestamp
+ */
+dashboard.clientState.sourceForLevel = function (scriptLevelId, timestamp) {
+  var data = localStorage.getItem('source' + scriptLevelId);
+  if (data) {
+    var parsed = JSON.parse(data);
+    if (!timestamp || parsed.timestamp > timestamp) {
```

Comment:
If parse doesn't have a timestamp this will always default to the local source; does that sound "right"?  That's a pathological case so maybe I shouldn't worry about it.